File: DSA/SakshamDwivedi_590013147/Day24/D24P3.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node* next;
};

int size = 0;
int m;
struct node* head = NULL;
void createNode(int);
void Print();
void Delete(int);

int main(){
    createNode(1);
    createNode(2);
    createNode(3);
    createNode(4);
    createNode(5);
    createNode(6);
    printf("Enter the position : ");
    scanf("%d", &m);
    printf("Before : ");
    Print();
    printf("\n");
    printf("After : ");
    Delete(m);
    Print();
}

void createNode(int data){
       struct node* Node = (struct node*)malloc(sizeof(struct node));
       Node -> data = data;
       Node -> next = NULL;
       if(head == NULL){
            head = Node;
            size++;
            return;
       }
     struct node* current = head;
     while(current -> next != NULL){
      current = current -> next;
     }
     current -> next = Node;
     size++;
}

void Print(){
    struct node* current = head;
    while(current != NULL){
        printf("%d -> ", current -> data);
        current = current -> next;
    }
    printf("NULL\n");

}

void Delete(int pos) {
    if(pos < 0 || pos > size){
        return;
    }
    struct node *current = head;
    struct node *freenode = head;
    if(pos == 1){
        head = current -> next;
        freenode = current;
        free(freenode);
        return;
    }
    for(int i = 1; i < pos - 1; i++){
        current = current -> next;
    }
    freenode = current -> next;
    current -> next = freenode -> next;
    free(freenode);
    size--;
}

File: DSA/SakshamDwivedi_590013147/Day24/D24P3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

size = 0
head = None

def create_node(data):
    global head, size
    new_node = Node(data)
    if head is None:
        head = new_node
        size += 1
        return
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    size += 1

def print_list():
    current = head
    while current:
        if current.next:
            print(current.data, " -> ", end="")
        else:
            print(current.data)
        current = current.next

def delete_position(pos):
    global head, size
    if pos <= 0 or pos > size:
        return
    current = head
    freenode = head
    if pos == 1:
        head = current.next
        del current
        size -= 1
        return
    i = 1
    while i < pos - 1:
        current = current.next
        i += 1
    freenode = current.next
    current.next = freenode.next
    del freenode
    size -= 1

create_node(1)
create_node(2)
create_node(3)
create_node(4)
create_node(5)
create_node(6)

m = int(input("Enter the position : "))

print("Before : ", end="")
print_list()

print("After : ", end="")
delete_position(m)
print_list()
